File: tests/python/math_operations.py
"""
Module containing deterministic mathematical operations.
"""
import math


class MathOperations:
    """Class containing deterministic mathematical operations."""

    def add(self, a: int, b: int) -> int:
        """Adds two integers."""
        return a + b

    def subtract(self, a: int, b: int) -> int:
        """Subtracts two integers."""
        return a - b

    def multiply(self, a: int, b: int) -> int:
        """Multiplies two integers."""
        return a * b

    def divide(self, a: int, b: int) -> float:
        """Divides two numbers. Raises ZeroDivisionError if divisor is zero."""
        if b == 0:
            raise ZeroDivisionError("Division by zero")
        return a / b

    def power(self, base: float, exponent: int) -> float:
        """Calculates the power of a number."""
        return math.pow(base, exponent)

    def factorial(self, n: int) -> int:
        """Calculates the factorial of a number. Raises ValueError if n is negative."""
        if n < 0:
            raise ValueError("Factorial is not defined for negative numbers")
        if n == 0 or n == 1:
            return 1
        result = 1
        for i in range(2, n + 1):
            result *= i
        return result

    def derivative(self, coefficients: list[float]) -> list[float]:
        """
        Computes the derivative of a polynomial represented by its coefficients.
        coefficients[0] + coefficients[1]*x + coefficients[2]*x^2 + ...
        """
        if coefficients is None:
            raise ValueError("Coefficients array must not be null")
        if len(coefficients) <= 1:
            return [0.0]
        result = []
        for i in range(1, len(coefficients)):
            result.append(coefficients[i] * i)
        return result

    def pi(self) -> str:
        """Returns the first 40 decimals of the pi constant."""
        return "3.1415926535897932384626433832795028841971"

    def gcd(self, a: int, b: int) -> int:
        """Calculates the greatest common divisor using Euclidean algorithm."""
        a = abs(a)
        b = abs(b)
        while b != 0:
            a, b = b, a % b
        return a

File: tests/typescript/mathOperations.ts
/**
 * Class containing deterministic mathematical operations.
 */
export class MathOperations {
  /**
   * Adds two integers.
   */
  add(a: number, b: number): number {
    return a + b;
  }

  /**
   * Subtracts two integers.
   */
  subtract(a: number, b: number): number {
    return a - b;
  }

  /**
   * Multiplies two integers.
   */
  multiply(a: number, b: number): number {
    return a * b;
  }

  /**
   * Divides two numbers.
   * @throws {Error} if divisor is zero
   */
  divide(a: number, b: number): number {
    if (b === 0) {
      throw new Error("Division by zero");
    }
    return a / b;
  }

  /**
   * Calculates the power of a number.
   */
  power(base: number, exponent: number): number {
    return Math.pow(base, exponent);
  }

  /**
   * Calculates the factorial of a number.
   * @throws {Error} if n is negative
   */
  factorial(n: number): number {
    if (n < 0) {
      throw new Error("Factorial is not defined for negative numbers");
    }
    if (n === 0 || n === 1) {
      return 1;
    }
    let result = 1;
    for (let i = 2; i <= n; i++) {
      result *= i;
    }
    return result;
  }

  /**
   * Computes the derivative of a polynomial represented by its coefficients.
   * coefficients[0] + coefficients[1]*x + coefficients[2]*x^2 + ...
   */
  derivative(coefficients: number[]): number[] {
    if (coefficients === null || coefficients === undefined) {
      throw new Error("Coefficients array must not be null");
    }
    if (coefficients.length <= 1) {
      return [0];
    }
    const result: number[] = [];
    for (let i = 1; i < coefficients.length; i++) {
      result.push(coefficients[i] * i);
    }
    return result;
  }

  /**
   * Returns the first 40 decimals of the pi constant.
   */
  pi(): string {
    return "3.1415926535897932384626433832795028841971";
  }

  /**
   * Calculates the greatest common divisor using Euclidean algorithm.
   */
  gcd(a: number, b: number): number {
    a = Math.abs(a);
    b = Math.abs(b);
    while (b !== 0) {
      const temp = b;
      b = a % b;
      a = temp;
    }
    return a;
  }
}
